### PR TITLE
Use json-memfd to pass "origin-ip" from -tls to -ws (enabling remote IP logging)

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
For a long time (since `cockpit-tls` landed) we've been unable to log the IP address of remote connections.   Use our already-existing memfd-json utilities to add a metadata blob passed from `cockpit-tls` to `cockpit-ws` as a way of getting the connection metadata (specifically, the remote IP) into `-ws`.

### CODE REVIEW

File: src/common/cockpitmemfdread.c [hunk 1]
```diff
@@ -137,3 +139,27 @@ cockpit_memfd_read_from_envvar (gchar      **result,
...
+  g_autofree gchar *content = cockpit_memfd_read (fd, error);
+
+  if (content == NULL)
+    return NULL;
+
+  return cockpit_json_parse_object (content, -1, error);
+}
+
+JsonObject *
+cockpit_memfd_read_json_from_control_messages (CockpitControlMessages  *ccm,
```
Comment (martinpitt) [inline src/common/cockpitmemfdread.c:156]: Please rebase your branch so that the commit that introduces `CockpitControlMessages` comes before this one ("memfdread: add a couple more public accessors"). Right now it's both confusing and non-bisectable.

File: src/common/cockpitcontrolmessages.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+{
+  if (ccm->n_messages == 0)
+    {
+      g_set_error (error, G_FILE_ERROR, G_FILE_ERROR_INVAL,
+                   "Unexpectedly received no control messages (one message of type %s expected)",
+                   g_type_name (message_type));
+      return NULL;
+    }
+
+  if (ccm->n_messages != 1)
```
Comment (martinpitt) [inline src/common/cockpitcontrolmessages.c:54]: Seems a bit overzealous to handle < 1 and > 1 differently..

File: src/common/cockpitcontrolmessages.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  ccm->n_messages = 0;
+}
+
+gboolean
+cockpit_control_messages_empty (CockpitControlMessages *ccm)
+{
+  return ccm->n_messages == 0;
+}
+
+gpointer
```
Comment (martinpitt) [inline src/common/cockpitcontrolmessages.c:41]: This currently always returns a `GSocketControlMessage*`. Why does it get anonymized to a typeless pointer? Should this perhaps unwrap the `GSocketControlMessage` and return its data instead?

File: src/common/cockpitcontrolmessages.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+  return ccm->messages[0];
+}
+
+const gint *
+cockpit_control_messages_peek_fd_list (CockpitControlMessages  *ccm,
+                                       gint                    *n_fds,
+                                       GError                 **error)
+{
+  GUnixFDMessage *message = cockpit_control_messages_get_single_message (ccm, G_TYPE_UNIX_FD_MESSAGE, error);
```
Comment (martinpitt) [inline src/common/cockpitcontrolmessages.c:78]: So here it does expect a `GSocketControlMessage*` again. As you are downcasting, can you please use a dynamic/checked glib cast?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Reviewed the first three commits, ran out of time. Will do the others later.

File: src/common/test-jsonfds.c [hunk 5]
```diff
@@ -0,0 +1,905 @@
...
+
+/* --- testing of printing --- */
+
+typedef struct
+{
+  char buffer[1024];
+  FILE *stream;
+
+  char expected_buffer[1024];
+  gchar *expected;
```
Comment (martinpitt) [inline src/common/test-jsonfds.c:44]: I find that naming a bit hard to follow when reading the code. `helpers->expected` pretty well always points to a `\0`, so it's not really the "expected" output (that is `expected_buffer`), more like an "expected_end".

File: src/common/test-jsonfds.c [hunk 6]
```diff
@@ -0,0 +1,905 @@
...
+helpers_compare_expected (Helpers *helpers)
+{
+#if 0
+  g_debug ("\n\n%.*s\n\nVS\n\n%.*s\n\n",
+           (int) ftell (helpers->stream), helpers->buffer,
+           (int) (helpers->expected - helpers->expected_buffer), helpers->expected_buffer);
+#endif
+
+  g_assert_cmpint (ftell (helpers->stream), ==, helpers->expected - helpers->expected_buffer);
+  g_assert (memcmp (helpers->buffer, helpers->expected_buffer, ftell (helpers->stream)) == 0);
```
Comment (martinpitt) [inline src/common/test-jsonfds.c:116]: Not very useful when it fails, but supposedly this is easier to look at in gdb than trying to print both? I thought everything non-printable was supposed to get escaped or replaced with `?`, so why can't this use `g_assert_cmpstr)`?

File: src/common/test-jsonfds.c [hunk 7]
```diff
@@ -0,0 +1,905 @@
...
+  return result;
+}
+
+static void
+test_string (Helpers    *helpers,
+             const char *key,
+             const char *value,
+             gint        length,
+             gint        min_length)
+{
```
Comment (martinpitt) [inline src/common/test-jsonfds.c:143]: Suggestion: add an assertion that `strlen(value) > length`, so that the test below is actually meaningful.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Very thorough test cases!

File: src/common/cockpitcontrolmessages.c (hunk 3 above)
Comment (allisonkarlitskaya) [reply src/common/cockpitcontrolmessages.c:41]: This could return any subclass of `GSocketControlMessage`, with error checking, as specified by the `message_type` field.  In particular, the function directly below calls it with `G_TYPE_UNIX_FD_MESSAGE` and expects to get a `GUnixFDMessage *` from it.

File: src/common/test-jsonfds.c [hunk 8]
```diff
@@ -0,0 +1,905 @@
...
+  char buffer[1024];
+  FILE *stream;
+
+  char expected_buffer[1024];
+  gchar *expected;
+
+  size_t pagesize;
+  char *accessible;
+  char *inaccessible;
+} Helpers;
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:49]: This is not exactly a `TestCase`, since it doesn't get initialised differently per test.  "Harness", I'd accept.

File: src/common/test-jsonfds.c [hunk 9]
```diff
@@ -0,0 +1,905 @@
...
+
+static void
+helpers_init (Helpers *helpers)
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:52]: This much I agree with.  Even as I was writing this, I was thinking about it.  I'll make this change.

File: src/common/test-jsonfds.c [hunk 10]
```diff
@@ -0,0 +1,905 @@
...
+{
+  /* set up the stream */
+  helpers->stream = fmemopen (helpers->buffer, sizeof helpers->buffer, "w");
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:55]: `tmpfile()` is another cool case of an unusual `FILE *` instance, and of course there's the all-powerful `fopencookie()` which lets you do basically anything you want.

`stdio` is underrated...

File: src/common/test-jsonfds.c (hunk 5 above)
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:44]: Reasonable proposal.  I'll change it.

File: src/common/test-jsonfds.c (hunk 6 above)
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:116]: Indeed.  In fact, this whole thing is a bit annoying.  What I want to be doing is checking case by case that the correct thing ends up in the stream.  What I actually end up doing is checking once at the end if the entire blob is OK, with absolutely no debug output (except the `#if 0`'d-out thing above) as to why it failed.

I'll rethink this.

File: src/common/test-jsonfds.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/common/test-jsonfds.c:49]: That's exactly how `TestCase` is being used everywhere. The "different initialization per test" should come from a `TestFixture` (optional)

[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:49]: Gotcha!  I don't write a lot of tests... does it show? :)

File: src/common/test-jsonfds.c (hunk 7 above)
Comment (allisonkarlitskaya) [reply src/common/test-jsonfds.c:143]: `length` is passed to be taken as the actual length of `value` (which is otherwise not nul-terminated)

File: src/common/cockpitcontrolmessages.c (hunk 4 above)
Comment (allisonkarlitskaya) [reply src/common/cockpitcontrolmessages.c:78]: The `GType` field given here is determining the return type of the function, so I consider this safe.

**Review by Gundersanne — APPROVED** [review]
I'm not sure how valuable it is* but this seems fine, and tests pass (retriggere dthem anyway to see if we can get something greener).

edit: * this refers to my review, not the code. The code makes sense, no obvious errors that I can see, and it's tested.

**Comment by allisonkarlitskaya** [comment]
> I'm not sure how valuable it is

We currently we have no logging of origin IP address for logins (or attempted logins).  This means we don't see the IP in `w`, `who`, `last`, or anywhere else.  It also means that any PAM module that applies policy based on the origin IP address (for example, blacklisting an IP range after too many failed login attemps) cannot work.  The main value of this PR is that we fix that.

Additionally, the metadata channel which we establish between `cockpit-tls` and `cockpit-ws` will be used in the future for more things: in particular, passing the "nonce" that we intend to use in our new approach to smartcard auth.

**Review by martinpitt — APPROVED** [review]
Great work, thank you! This looks very clean to me now, and my previous comments all got addressed (except for the `return gpointer` thing, which I still don't like, but also won't make a big fuss about :grin: )

Nice to have this back! I add release-note to this, as this is totally worth mentioning.

**Comment by allisonkarlitskaya** [comment]
for release notes:

Logging of remote IP addresses

Cockpit now includes the remote IP address when logging user logins (and login attempts).  This means that the IP address will appear in the system logs, in the output of tools such as `w` or `who`, and in the login-time messages about when the last successful login was, and any failed login attempts.  It also means that PAM modules can now filter logins based on IP address (for example, temporarily blocking a subnet when there are too many failed login attempts).
